src/data_source/storage/sql.rs

```diff
@@ -3147,6 +3148,16 @@ pub mod testing {
                 tracing::warn!("database is not ready");
                 sleep(Duration::from_secs(1)).await;
             }
+
+            // The above command ensures the database is ready inside the Docker container. However,
+            // on some systems, there is a slight delay before the port is exposed via host
+            // networking. We don't need to check again that the database is ready on the host (and
+            // maybe can't, because the host might not have pg_isready installed), but we can
+            // ensure the port is open by just establishing a TCP connection.
+            while let Err(err) = TcpStream::connect(format!("{}:{}", self.host, self.port)).await {
+                tracing::warn!("database is ready, but port is not available to host: {err:#}");
+                sleep(Duration::from_secs(1)).await;
```

Comment:
Should we consider adding a timeout?

Reply:
I'm torn. Realistically the only thing we could do after a timeout is panic...so it seems better to keep retrying and let the test runner worry about timeouts. OTOH this could be annoying for local testing. Maybe a timeout that is configurable via an env var so we can easily adjust it if it is just set too short for CI or anything?

Reply:
[be8de83](https://github.com/EspressoSystems/hotshot-query-service/pull/655/commits/be8de83e07327a96d494a708c73336c148445846)